shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -360,78 +427,190 @@
+  auto zIndex = 0;
   for (size_t i = 0; i < composition_order_.size(); i++) {
-    int view_id = composition_order_[i];
-    // We added a chain of super views to the platform view to handle clipping.
-    // The `platform_view_root` is the view at the top of the chain which is a direct subview of the
-    // `FlutterView`.
-    UIView* platform_view_root = root_views_[view_id].get();
-    UIView* overlay = overlays_[view_id]->overlay_view;
-    FML_CHECK(platform_view_root.superview == overlay.superview);
-    if (platform_view_root.superview == flutter_view) {
-      [flutter_view bringSubviewToFront:platform_view_root];
-      [flutter_view bringSubviewToFront:overlay];
-    } else {
+    auto platform_view_id = composition_order_[i];
+    auto layers = layer_map[platform_view_id];
+    auto platform_view_root = root_views_[platform_view_id].get();
+
+    if (platform_view_root.superview != flutter_view) {
       [flutter_view addSubview:platform_view_root];
-      [flutter_view addSubview:overlay];
-      overlay.frame = flutter_view.bounds;
+    } else {
+      platform_view_root.layer.zPosition = zIndex++;
     }
-
-    active_composition_order_.push_back(view_id);
+    for (auto layer : layers) {
+      if ([layer->overlay_view superview] != flutter_view) {
+        [flutter_view addSubview:layer->overlay_view];
+      } else {
+        layer->overlay_view.get().layer.zPosition = zIndex++;
+      }
+    }
+    active_composition_order_.push_back(platform_view_id);
   }
-  composition_order_.clear();
-  return did_submit;
 }
 
-void FlutterPlatformViewsController::DetachUnusedLayers() {
+std::shared_ptr<FlutterPlatformViewLayer> FlutterPlatformViewsController::GetLayer(
+    GrContext* gr_context,
+    std::shared_ptr<IOSContext> ios_context,
+    sk_sp<SkPicture> picture,
+    SkRect rect,
+    int64_t view_id,
+    int64_t overlay_id) {
+  auto layer = layer_pool_->GetLayer(gr_context, ios_context);
+  auto screenScale = [UIScreen mainScreen].scale;
+  // Set the size of the overlay UIView.
+  layer->overlay_view.get().frame = CGRectMake(rect.x() / screenScale,      //
+                                               rect.y() / screenScale,      //
+                                               rect.width() / screenScale,  //
+                                               rect.height() / screenScale  //
+  );
+  // Set a unique view identifier, so the overlay can be identified in unit tests.
+  layer->overlay_view.get().accessibilityIdentifier =
+      [NSString stringWithFormat:@"platform_view[%lld].overlay[%lld]", view_id, overlay_id];
+
+  std::unique_ptr<SurfaceFrame> frame =
+      layer->surface->AcquireFrame(SkISize::Make(rect.width(), rect.height()));
+  // If frame is null, AcquireFrame already printed out an error message.
+  if (frame == nullptr) {
+    return layer;
+  }
+  auto overlay_canvas = frame->SkiaCanvas();
+  overlay_canvas->clear(SK_ColorTRANSPARENT);
+  overlay_canvas->translate(-rect.x(), -rect.y());
```

Comment:
Maybe we should include a comment here for future reference for why we added this translate?